### PR TITLE
devenv: allow for python 3.8, and prepare CI configuration for dual 3.6/3.8

### PR DESCRIPTION
**Pull request by joshuarli**
People can now set `SENTRY_PYTHON_VERSION=3.8.12` in their `.env`, and follow what direnv has to say in order to set up a new 3.8 venv. For example if I have a 3.6.10 venv and do direnv allow:

```
ERROR: You have explicitly set a non-recommended Python version (3.8.10),
but it doesn't match the value of python's version: 3.6.13
You should create a new 3.8.10 virtualenv by running  "rm -rf .venv && direnv allow".
```

Then, we'll build+install 3.8.10 if we don't have it already, and finish bootstrapping the new venv.

I'm gonna follow up with getsentry, and then after that, a PR that adds required tests on 3.8 in addition to 3.6, and sending out a mail on how to move devenv to 3.8. It'll be pretty expensive, but we need to ensure dual compatibility during this transitional period. I won't merge that PR until a full 3.8 deploy is initiated, and after things look good I'll put up a PR to remove 3.6 entirely from devenv and tests.

### CODE REVIEW

File: scripts/lib.sh
Comment (evanpurkhiser) [inline scripts/lib.sh:69]: Why Why does this have the whitespace at the front?

File: scripts/lib.sh
Comment (joshuarli) [inline scripts/lib.sh:65]: I moved some stuff from ensure-venv to here. To me, it makes more sense to print those validation errors in the function itself rather than in a specific caller (ensure-venv).

File: scripts/lib.sh
Comment (joshuarli) [reply scripts/lib.sh:69]: lol

File: scripts/lib.sh
Comment (evanpurkhiser) [inline scripts/lib.sh:73]: this one too josh

File: scripts/lib.sh
Comment (evanpurkhiser) [inline scripts/lib.sh:59]: Will this mean anyone currently using M1 (I assume basically no one) will have to rebuild their venv?

Comment (joshuarli) [reply scripts/lib.sh:59]: Yeah. I'm targetting 3.8.12.

**Comment by joshuarli** [comment]
huh `Copying src/sentry.egg-info to /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/sentry-21.10.0.dev0-py3.6.egg-info` vs failing 

```
Copying src/sentry.egg-info to /usr/local/lib/python3.8/dist-packages/sentry-21.10.0.dev0.egg-info
error: [Errno 13] Permission denied: '/usr/local/lib/python3.8/dist-packages/sentry-21.10.0.dev0.egg-info'
```

maybe i do need a few CI changes here as well

**Comment by joshuarli** [comment]
Oh yeah, this will just be testing on 3.8 by default right now. I gotta remain on 3.6.10.

**Review by evanpurkhiser — DISMISSED** [review]
lgtm

**Review by billyvg — COMMENTED** [review]
We'll need to update the GH branch settings before merging - can you actually cherry-pick these changes so that we can limit the number of times we have to update the branch settings? https://github.com/getsentry/sentry/pull/29042/files

I think we'll want `timeout-minutes: 20` instead of `15` as the latter is a bit aggressive.

**Comment by joshuarli** [comment]
Oh dang, this is definitely due to psycopg-binary.

```
tests/sentry/api/endpoints/test_user_notification_fine_tuning.py RRRRRER [ 24%]
##[debug]Re-evaluate condition on job cancellation for step: 'Run backend test (1 of 4)'.
Error: The operation was canceled.
```

**Comment by joshuarli** [comment]
d7a4765 has been moved to https://github.com/getsentry/sentry/pull/29064.

File: .github/workflows/acceptance.yml
Comment (joshuarli) [inline .github/workflows/acceptance.yml:151]: I think this might mess up the freight stuff since the check titles change as a result? Will have to take a look at it before merging, I might just modify it to ignore stuff enclosed in `()` or something.

Comment (joshuarli) [reply .github/workflows/acceptance.yml:151]: Oh well, this is a getsentry concern, so doesn't block this PR, but what about the cross repo machinery @billyvg?

File: .github/file-filters.yml
Comment (armenzg) [inline .github/file-filters.yml:46]: Good eye

File: scripts/lib.sh
Comment (armenzg) [reply scripts/lib.sh:59]: Hey, hey! Not no one! I'm still here, everyone! :)

wfm

File: scripts/pyenv_setup.sh
Comment (armenzg) [inline scripts/pyenv_setup.sh:99]: Is it only `3.6.10` that needs the code patching?

**Review by armenzg — APPROVED** [review]
lgtm

File: .github/workflows/acceptance.yml
Comment (billyvg) [reply .github/workflows/acceptance.yml:151]: @joshuarli hmm, it shouldn't be affected for cross-repo tasks, but yeah it will affect freight if we do the same things in getsentry.

**Comment by joshuarli** [comment]
Oh @billyvg I think the repo settings need to be changed too? I don't think this can be avoided unfortunately, we'll have to do this when I add 3.8 too.

`acceptance / acceptance (0, 3.6.13) (pull_request) Successful in 12m `
`acceptance (0) Expected — Waiting for status to be reported (Required)`

**Pull request merged**
